### DISCUSSION TITLE
How to track the mouse drag over interactive images?

### DISCUSSION THREAD

**Discussion by smojef** [zauberzeug/nicegui#425] [Q&A] [answered]
Almost there, but I did not find how to detect mouse release outside of the image.
```python
from nicegui import ui
from nicegui.events import MouseEventArguments

down = False
def mouse_handler(e: MouseEventArguments):
    global down
    if e.type == 'mousedown':
        down = True
    if e.type == 'mouseup':
        down = False
    #if e.type in ('mouseleave', 'pointercancel'):
    #    down = False  # Drag will be lost if pointer leaves image (even if it comes back inside).
    if e.type in ('mousedown', 'mousemove') and down:
        ii.content = f'<circle cx="{e.image_x}" cy="{e.image_y}" r="10" fill="Red" />'
    mylog.push(f'{e.type} at ({e.image_x:.1f}, {e.image_y:.1f})')

src = 'https://picsum.photos/id/565/640/360'
ii = ui.interactive_image(src, on_mouse=mouse_handler, events=['mousedown', 'mousemove', 'mouseup', 'mouseleave', 'pointerdown', 'pointerup', 'pointermove', 'pointercancel', 'gotpointercapture'], cross=True)
ui.button('info', on_click=lambda _: ui.notify(str(ii.content)))
mylog = ui.log(200).classes('w-full h-96')
ui.run(port=9000)
```
I tried watching many possible events.
The code above does not stop the mouse drag if the mouse button is released outside of the image.
When uncommenting the `if`, then it loses the mouse drag if we leave the image (even without releasing the mouse button).
The goal was to use the interactive image as an X, Y, or X+Y input control (with background reference).
But we need a way to handle boundaries.

**Comment by falkoschindler** [comment]
Interesting problem. You should be able to distinguish the re-entry with or without the mouse button being pressed. Unfortunately, mouse buttons are not part of the `MouseEventArguments`. I experimented with a second event handler:

```py
def move_handler(msg: dict):
    global down
    if msg['args']['buttons'] == 0:
        down = False

ii.on('mousemove', move_handler, args=['buttons'])
```

However, this handler is called _after_ your `mouse_handler`, so the content is redrawn before updating `down`.

A cleaner solution would probably be to add more information to the `MouseEventArguments`, like buttons and modifier keys. Then you could process all cases in one handler.

But you aren't trying to re-implement a `ui.joystick`, are you?

**Reply by smojef** [reply]
No, `ui.joystick` serves a different purpose, by tracking relative movements.

I want to allow a user to put a pin on a map (image) and be able to move it around so the system can react to the coordinates. Like providing a heat map to cue the user about the system reaction.

It all worked well until I tested draging outside the image.

Isn't there a way to 'capture' the mouse pointer on mouse down so that it is notified of mouse up from outside?

But I agree, having the mouse button status in the same event could solve this.

**Reply by smojef** [reply]
Other thoughts:
- How come your `move_handler` does not provide coordinates?
- Is there any other nicegui elements that handle mouse dragging events?

**Comment by falkoschindler** [comment]
> No, ui.joystick serves a different purpose, by tracking relative movements.
> I want to allow a user to put a pin on a map (image) and be able to move it around so the system can react to the coordinates. Like providing a heat map to cue the user about the system reaction.
I see.

> Isn't there a way to 'capture' the mouse pointer on mouse down so that it is notified of mouse up from outside?
Not that I'm aware of.

> How come your `move_handler` does not provide coordinates?
The idea of `interactive_image` is to work in image coordinates (independent of the rendered size). Normal mouse events (e.g. my `move_handler`) yield different coordinates, but not relative to the original image size. But the `MouseEventArguments` contain such "image coordinates".

> Is there any other nicegui elements that handle mouse dragging events?
Basically every element supports dragging events. See #397 for a related discussion.

I think I'll open an issue for extending the `MouseEventArguments` with buttons and modifier keys.

**Reply by smojef** [reply]
Yes I definitely need the buttons modifiers in `MouseEventArguments` in order to obtain dragging coordinates correctly.

**Reply by smojef** [reply]
FYI, I tried hacking the `ui.joystick`.  I can obtain absolute coordinates, but can't hide the nipple, change the background, its size or its aspect ratio:
```python
from nicegui import ui

with ui.row():
    ui.label('---------------')
    ui.joystick(
        mode='static', 
        restJoystick=False, 
        restOpacity=1, 
        threshold=0.01,
        color='blue', 
        shape='square', 
        size=200, # Not handled well and does not allow to change aspect ratio
        #dataOnly=True, # does not work
        #position=dict(top='50px', left='10px', bottom='100px', right='500px'), # still square
        dynamicPage=True,
        on_move=lambda msg: mylog.push(str(msg)),
    ) #.classes('w-96') # Does not work
# We would then need to hide the nipple and overlay (under?) an interactive image with a circle at the coordinates
mylog = ui.log(200).classes('w-full h-96')
ui.run(port=9000)
```

I also tried the trello example from #397, but I cannot find coordinates in any events.

**Reply by falkoschindler** [reply]
Yeah, nipplejs doesn't seem to be the way to go. However, I'm wondering how they implement the dragging, which doesn't stop when leaving the container area until the mouse button is released...

**Reply by smojef** [reply]
I was once familiar with the concept of "capturing" the mouse pointer on mouse_down so that the element keeps receiving the events.

Have a look at this reference: https://developer.mozilla.org/en-US/docs/Web/API/Element/setPointerCapture
The example in the Result box is exactly what I try to achieve.

**Comment by falkoschindler** [comment, marked as answer]
With the recent commit https://github.com/zauberzeug/nicegui/commit/cbdeffad14b77a745c1aa7f93d47fdeaa85754af the following example works by analyzing the mouse buttons within the mouse event handler:

```py
#!/usr/bin/env python3
from nicegui import ui
from nicegui.events import MouseEventArguments

down = False

def mouse_handler(e: MouseEventArguments):
    global down
    if e.type == 'mousedown':
        down = True
    if e.type == 'mouseup':
        down = False
    if e.buttons == 0:
        down = False
    if down:
        ii.content = f'<circle cx="{e.image_x}" cy="{e.image_y}" r="10" fill="Red" />'

src = 'https://picsum.photos/id/565/640/360'
ii = ui.interactive_image(src, on_mouse=mouse_handler, events=['mousedown', 'mousemove', 'mouseup'], cross=True)

ui.run()
```

---

You can also use pointer capturing. Thanks for the hint! I didn't know about `setPointerCapture`. Note that we need to capture the pointer on the first child, which is the `img` tag with all the event listeners.

```py
#!/usr/bin/env python3
from nicegui import ui
from nicegui.events import MouseEventArguments

down = False

def mouse_handler(e: MouseEventArguments):
    if down:
        ii.content = f'<circle cx="{e.image_x}" cy="{e.image_y}" r="10" fill="Red" />'

async def pointerdown(msg: dict):
    global down
    down = True
    pointer_id = msg['args']['pointerId']
    await ui.run_javascript(f'document.getElementById("{ii.id}").firstChild.setPointerCapture({pointer_id})', respond=False)

async def pointerup(msg: dict):
    global down
    down = False
    # NOTE: pointer capture is released automatically on pointerup

src = 'https://picsum.photos/id/565/640/360'
ii = ui.interactive_image(src, on_mouse=mouse_handler, events=['mousemove'], cross=True)
ii.on('pointerdown', pointerdown, args=['pointerId'])
ii.on('pointerup', pointerup, args=['pointerId'])

ui.run(port=1234, uvicorn_reload_includes='*.py,*.js,*.css,*.html')
```

**Reply by smojef** [reply]
Great! This is amazing!
